pkg/services/statefetcher/statefetcher.go

```diff
@@ -0,0 +1,407 @@
+	for h := currentHeight; h > 0; h-- {
+		root, err := bfs.stateModule.GetStateRoot(h)
+		if err == nil {
+			lastSavedRoot = root
+			lastSavedHeight = h
+			break
+		}
+	}
+	if lastSavedRoot != nil {
+		bfs.log.Info("found last saved state root in store",
+			zap.String("root", lastSavedRoot.Root.String()),
+			zap.Uint32("height", lastSavedHeight))
+	} else {
+		bfs.log.Info("no saved state root found in store up to current height",
+			zap.Uint32("currentHeight", currentHeight))
+	}
+	if lastSavedHeight == currentHeight-1 {
+		bfs.log.Info("no new state root to fetch, exiting")
+		return nil
+	}
+	reader, err := bfs.getMPTNodes()
+	if err != nil {
+		bfs.log.Error("failed to get MPT nodes", zap.Error(err))
+		return err
+	}
+	defer reader.Close()
+
+	br := gio.NewBinReaderFromIO(reader)
+
+	version := br.ReadB()
+	if version != 0 || br.Err != nil {
+		return fmt.Errorf("invalid version: %d, error: %w", version, br.Err)
+	}
+	_ = br.ReadU32LE()
+	syncHeight := br.ReadU32LE()
+	var stateRoot util.Uint256
+	br.ReadBytes(stateRoot[:])
+	if br.Err != nil {
+		return fmt.Errorf("failed to read NeoFS header: %w", br.Err)
+	}
+	if syncHeight > currentHeight {
+		return fmt.Errorf("sync height %d exceeds current height %d", syncHeight, currentHeight)
+	}
+
+	bfs.log.Info("NeoFS state root", zap.String("root", stateRoot.String()), zap.Uint32("height", syncHeight))
+
+	localMPT := mpt.NewTrie(mpt.EmptyNode{}, mpt.ModeAll, bfs.stateModule.Store)
+	bfs.log.Info("initial MPT root", zap.String("root", localMPT.StateRoot().String()))
+	batch := make(map[string][]byte, batchSize)
+	processed := 0
+	cache := storage.NewMemCachedStore(bfs.stateModule.Store)
+	tempPrefix := []byte{byte(storage.STTempStorage)}
+
+	for {
+		neoFSKey := br.ReadVarBytes()
+		if br.Err != nil {
+			if errors.Is(br.Err, io.EOF) {
+				break
+			}
+			bfs.log.Error("failed to read key", zap.Error(br.Err))
+			return br.Err
+		}
+		value := br.ReadVarBytes()
+		if br.Err != nil {
+			bfs.log.Error("failed to read value", zap.Error(br.Err))
+			return br.Err
+		}
+
+		fullKey := make([]byte, len(tempPrefix)+len(neoFSKey))
+		copy(fullKey, tempPrefix)
+		copy(fullKey[len(tempPrefix):], neoFSKey)
+		key := neoFSKey
+		if processed < 5 {
+			bfs.log.Debug("key sample", zap.ByteString("fullKey", fullKey), zap.ByteString("mptKey", key), zap.ByteString("value", value))
+		}
+
+		cache.Put(fullKey, value)
+		batch[string(key)] = value
+		processed++
+
+		if len(batch) >= batchSize {
+			if err := bfs.processMPTBatch(localMPT, batch, syncHeight, cache); err != nil {
+				bfs.log.Error("failed to process MPT batch", zap.Error(err))
+				return err
+			}
+			if n, err := cache.PersistSync(); err != nil {
+				bfs.log.Error("failed to persist cache", zap.Error(err), zap.Int("items", n))
+				return err
+			}
+			bfs.log.Info("batch root", zap.String("root", localMPT.StateRoot().String()))
+			batch = make(map[string][]byte, batchSize)
+			cache = storage.NewMemCachedStore(bfs.stateModule.Store)
```

Comment:
NACK. You should flush changes from memcached storage to the underlying store and then reuse the same memcached storage for further operations.